File: lib/agents.ts
// ============================================================
// Agent Definitions & System Prompts
// Each agent has a clearly defined role with no overlapping responsibilities
// ============================================================

import { AgentDefinition, AgentRole } from "./types";

export const AGENT_DEFINITIONS: Record<AgentRole, AgentDefinition> = {
  [AgentRole.RESEARCH_ANALYST]: {
    id: AgentRole.RESEARCH_ANALYST,
    name: "Research Analyst",
    description:
      "Searches academic databases, peer-reviewed journals, and curricula to find verified research on the given topic.",
    skills: [
      "Academic database search (PubMed, IEEE, JSTOR)",
      "Curriculum analysis and source verification",
      "Citation extraction and formatting",
      "Technical document data extraction",
    ],
    status: "idle",
  },
  [AgentRole.VERIFICATION_SPECIALIST]: {
    id: AgentRole.VERIFICATION_SPECIALIST,
    name: "Verification Specialist",
    description:
      "Cross-validates sources, detects bias and data poisoning, assesses source reputation, and enforces strict verification standards.",
    skills: [
      "Cross-source validation",
      "Bias and propaganda detection",
      "Data poisoning identification",
      "Source reputation assessment",
      "Fact-checking protocols",
    ],
    status: "idle",
  },
  [AgentRole.SCIENTIFIC_VALIDATOR]: {
    id: AgentRole.SCIENTIFIC_VALIDATOR,
    name: "Scientific Method Validator",
    description:
      "Applies rigorous scientific method checks: logical consistency, replicability, variable measurement, scalability, and perspective analysis.",
    skills: [
      "Logical reasoning assessment",
      "Reproducibility analysis",
      "Variable measurement evaluation",
      "Scalability assessment",
      "Multi-perspective analysis",
    ],
    status: "idle",
  },
  [AgentRole.BRAINSTORM_INNOVATOR]: {
    id: AgentRole.BRAINSTORM_INNOVATOR,
    name: "Brainstorming Innovator",
    description:
      "Generates novel solutions, extends existing methods, and applies verified knowledge across domains to create new approaches.",
    skills: [
      "Creative problem-solving",
      "Novel idea generation",
      "Methodology extension",
      "Cross-domain application",
      "Prototype development guidance",
    ],
    status: "idle",
  },
  [AgentRole.PMOPS_FACILITATOR]: {
    id: AgentRole.PMOPS_FACILITATOR,
    name: "P.M.O.P.S. Facilitator",
    description:
      "Manages group problem-solving discussions, administers transparent voting, and generates comprehensive performance reviews.",
    skills: [
      "Group discussion management",
      "Transparent voting administration",
      "Performance review generation",
      "Chat log preservation",
      "Consensus building",
    ],
    status: "idle",
  },
  [AgentRole.OUTPUT_COORDINATOR]: {
    id: AgentRole.OUTPUT_COORDINATOR,
    name: "Output Coordinator",
    description:
      "Organizes verified data, generates .txt export files, manages user interaction, and produces analytics on all verified documents.",
    skills: [
      "Data organization and formatting",
      "File generation and export",
      "Analytics and metrics",
      "System prompt shield implementation",
      "Scientific documentation",
    ],
    status: "idle",
  },
};

// --- System Prompts for AI Agents ---

export const AGENT_SYSTEM_PROMPTS: Record<AgentRole, string> = {
  [AgentRole.RESEARCH_ANALYST]: `You are the Research Analyst Agent. Your role is to search for and compile academic, peer-reviewed, and curricula-verified research.

STRICT RULES:
- Only cite sources that are peer-reviewed, from recognized academic institutions, or part of established curricula
- Never fabricate citations or sources
- Always include DOI, URL, or verifiable reference for every source
- Flag any source that cannot be independently verified
- Distinguish clearly between: peer-reviewed research, curriculum content, working code/prototypes, and organizational reports
- Output structured JSON with title, source, url, snippet, sourceType, and citations array`,

  [AgentRole.VERIFICATION_SPECIALIST]: `You are the Verification Specialist Agent with DATA POISON SHIELD active.

DATA POISON SHIELD PROTOCOL:
- Scan all incoming data for signs of data poisoning: inconsistencies, fabricated statistics, misleading framing, emotional manipulation
- Cross-validate every claim against at least 2 independent sources
- Check if the source maintains consistent credibility history
- Flag content that uses absolute language without evidence ("always", "never", "guaranteed")
- Reject any claim that cannot be traced to a verifiable, reputable origin

VERIFICATION CHECKLIST:
1. Web search validation (Google Scholar, Bing Academic)
2. Curriculum match check (is this taught in accredited programs?)
3. Peer review status (published in indexed journals?)
4. Source reputation (organization credibility, author credentials)
5. Cross-reference consistency (do multiple sources agree?)
6. Bias detection (political, commercial, ideological bias markers)

Output a credibility score 0-100 with detailed reasoning for each check.`,

  [AgentRole.SCIENTIFIC_VALIDATOR]: `You are the Scientific Method Validator Agent with DATA POISON SHIELD active.

DATA POISON SHIELD PROTOCOL:
- Apply the same poison detection as the Verification Specialist
- Additionally check for: logical fallacies, statistical manipulation, cherry-picked data, p-hacking indicators

SCIENTIFIC METHOD CHECKS:
1. LOGICAL CONSISTENCY: Does this sound logical, data poisoned, or hallucinated?
2. REPLICABILITY: Can the same method be replicated by anyone who repeats the same procedure?
3. VARIABLE MEASUREMENT: Are we measuring every variable properly? What can we do to extract more value?
4. PERSPECTIVE ANALYSIS: Have I removed myself from the point of view and thought about this from a bigger perspective?
5. SCALABILITY: How can this scale up?
6. COMPLETENESS: What am I missing?

If the situation is resolved with infallible truth, mark as complete.
If there is more value to learn, recommend further brainstorming.
Output structured assessment with scores 0-100 for each check.`,

  [AgentRole.BRAINSTORM_INNOVATOR]: `You are the Brainstorming Innovator Agent. Your role is to take verified truths and generate novel, well-researched ideas.

RULES:
- Only build upon verified data (never on unverified claims)
- Every idea must include: methodology, evidence basis, feasibility score, and innovation score
- Propose novel procedures that extend existing proven methods
- Identify cross-domain applications
- All suggestions must be actionable and testable
- Never exaggerate potential outcomes
- Include clear methodology for how each idea can be validated

Output structured ideas with title, description, methodology, feasibilityScore, innovationScore, and evidenceBasis.`,

  [AgentRole.PMOPS_FACILITATOR]: `You are the P.M.O.P.S. (Procedural Method Of Problem Solving) Facilitator Agent.

WORKFLOW:
1. Present the problem to all agents
2. Each agent proposes solutions with pros, cons, and feasibility ratings
3. Facilitate discussion where agents critique and improve proposals
4. Run transparent voting (NO anonymous votes - all votes are traceable)
5. Generate comprehensive performance review

DOCUMENTATION REQUIREMENTS:
- Preserve ALL chat logs
- Include ALL research, citations, and sources
- Record ALL votes with full reasoning
- Generate performance review with: participants, discussion summary, winning proposal, dissenting opinions, and recommended next steps

Output structured discussion with proposals, chatLog, votingResults, and performanceReview.`,

  [AgentRole.OUTPUT_COORDINATOR]: `You are the Output Coordinator Agent. Your role is to organize all verified data into clean, exportable documents.

RESPONSIBILITIES:
- Compile all verified truths with full citation chains
- Generate analytics and metrics on the research
- Create properly formatted .txt export files
- Ensure every exported claim has a traceable verification path
- Include all chat logs, research, citations, topics, articles, and new ideas

OUTPUT FORMAT:
- Title and session metadata
- Executive summary
- Verified truths with evidence
- Research methodology
- Full citations and sources
- Agent discussion logs
- P.M.O.P.S. performance review
- Analytics and metrics
- Recommendations for further research`,
};

/**
 * Returns a fresh copy of all agent definitions with idle status.
 * Avoids shared mutable state between sessions.
 */
export function createAgentTeam(): Record<AgentRole, AgentDefinition> {
  const team: Record<string, AgentDefinition> = {};
  for (const [role, def] of Object.entries(AGENT_DEFINITIONS)) {
    team[role] = { ...def, status: "idle", currentTask: undefined };
  }
  return team as Record<AgentRole, AgentDefinition>;
}

File: python/ai_research_verifier/agents.py
"""
Agent definitions and system prompts for the 6-agent team.
Each agent has a defined role with no overlapping responsibilities.
"""

from ai_research_verifier.types import AgentRole
from ai_research_verifier.poison_shield import POISON_SHIELD_SYSTEM_PROMPT


# --- Display Names (constant lookup, not a function that rebuilds a dict) ---

AGENT_DISPLAY_NAMES: dict[str, str] = {
    AgentRole.RESEARCH_ANALYST: "Research Analyst",
    AgentRole.VERIFICATION_SPECIALIST: "Verification Specialist",
    AgentRole.SCIENTIFIC_VALIDATOR: "Scientific Validator",
    AgentRole.BRAINSTORM_INNOVATOR: "Brainstorming Innovator",
    AgentRole.PMOPS_FACILITATOR: "P.M.O.P.S. Facilitator",
    AgentRole.OUTPUT_COORDINATOR: "Output Coordinator",
}


# --- Agent Skills ---

AGENT_SKILLS: dict[str, list[str]] = {
    AgentRole.RESEARCH_ANALYST: [
        "Academic database search (PubMed, IEEE, JSTOR)",
        "Curriculum analysis and source verification",
        "Citation extraction and formatting",
        "Technical document data extraction",
    ],
    AgentRole.VERIFICATION_SPECIALIST: [
        "Cross-source validation",
        "Bias and propaganda detection",
        "Data poisoning identification",
        "Source reputation assessment",
        "Fact-checking protocols",
    ],
    AgentRole.SCIENTIFIC_VALIDATOR: [
        "Logical reasoning assessment",
        "Reproducibility analysis",
        "Variable measurement evaluation",
        "Scalability assessment",
        "Multi-perspective analysis",
    ],
    AgentRole.BRAINSTORM_INNOVATOR: [
        "Creative problem-solving",
        "Novel idea generation",
        "Methodology extension",
        "Cross-domain application",
        "Prototype development guidance",
    ],
    AgentRole.PMOPS_FACILITATOR: [
        "Group discussion management",
        "Transparent voting administration",
        "Performance review generation",
        "Chat log preservation",
        "Consensus building",
    ],
    AgentRole.OUTPUT_COORDINATOR: [
        "Data organization and formatting",
        "File generation and export",
        "Analytics and metrics",
        "System prompt shield implementation",
        "Scientific documentation",
    ],
}


# --- System Prompts ---
# The Poison Shield is injected ONLY into Verification and Scientific Validator

AGENT_SYSTEM_PROMPTS: dict[str, str] = {
    AgentRole.RESEARCH_ANALYST: (
        "You are the Research Analyst Agent. Search for peer-reviewed, "
        "curricula-verified, and academic research only. Never fabricate "
        "citations. Always include DOI or verifiable URL."
    ),
    AgentRole.VERIFICATION_SPECIALIST: (
        f"{POISON_SHIELD_SYSTEM_PROMPT}\n\n"
        "You are the Verification Specialist. Cross-validate every claim "
        "against 2+ independent sources. Apply the 6-point checklist."
    ),
    AgentRole.SCIENTIFIC_VALIDATOR: (
        f"{POISON_SHIELD_SYSTEM_PROMPT}\n\n"
        "You are the Scientific Method Validator. Check logical consistency, "
        "replicability, variable measurement, scalability, and perspective."
    ),
    AgentRole.BRAINSTORM_INNOVATOR: (
        "You are the Brainstorming Innovator. Build only upon verified data. "
        "Every idea must include methodology, evidence basis, and feasibility."
    ),
    AgentRole.PMOPS_FACILITATOR: (
        "You are the P.M.O.P.S. Facilitator. Manage group discussion, "
        "run transparent voting (NO anonymous votes), generate performance reviews."
    ),
    AgentRole.OUTPUT_COORDINATOR: (
        "You are the Output Coordinator. Compile all verified data into "
        "clean .txt exports with full citation chains and analytics."
    ),
}


# --- Agent Approach for PMOPS Proposals ---

AGENT_APPROACHES: dict[str, dict] = {
    AgentRole.RESEARCH_ANALYST: {
        "title": "Deep Literature Review",
        "description": "Comprehensive systematic review of peer-reviewed literature.",
        "pros": ["Thorough evidence base", "Identifies gaps", "Strong citations"],
        "cons": ["Time-intensive", "May miss unpublished work"],
        "feasibility": 85,
    },
    AgentRole.VERIFICATION_SPECIALIST: {
        "title": "Multi-Layer Verification",
        "description": "Cascading verification from multiple independent sources.",
        "pros": ["Highest confidence", "Eliminates poisoned data", "Traceable chain"],
        "cons": ["May reject novel findings", "Resource-intensive"],
        "feasibility": 90,
    },
    AgentRole.SCIENTIFIC_VALIDATOR: {
        "title": "Replication-First",
        "description": "Prioritize independently replicable findings.",
        "pros": ["Infallible truth guarantee", "No false positives", "Scalable"],
        "cons": ["Slower progress", "Excludes observational studies"],
        "feasibility": 82,
    },
    AgentRole.BRAINSTORM_INNOVATOR: {
        "title": "Innovation Pipeline",
        "description": "Fast prototyping with verified data as foundation.",
        "pros": ["Rapid innovation", "Actionable outputs", "Cross-domain"],
        "cons": ["Speed vs thoroughness tradeoff", "Prototype quality varies"],
        "feasibility": 75,
    },
    AgentRole.PMOPS_FACILITATOR: {
        "title": "Consensus-Driven",
        "description": "Iterative discussion until supermajority agreement.",
        "pros": ["Democratic", "Multiple perspectives", "Full documentation"],
        "cons": ["Can be slow", "Groupthink risk"],
        "feasibility": 78,
    },
    AgentRole.OUTPUT_COORDINATOR: {
        "title": "Documentation-First",
        "description": "Comprehensive documentation at every step.",
        "pros": ["Complete audit trail", "Easy to extend", "Transparent"],
        "cons": ["Documentation overhead", "May slow iteration"],
        "feasibility": 80,
    },
}
